Fixes for tracking struct field sequences

briansull opened the pull request:
Fixes for Zero Offset field sequence tracking

 - A GT_LCL_VAR may have a zeroOffset field
 - Add an assert to prevent building field sequences with duplicates
 - Fix fgMorphField when we have a zero offset field

Improve fgAddFieldSeqForZeroOffset
 - Add JItDump info
 - Handle GT_LCL_FLD

briansull commented:
@dotnet-build-bot Test Ubuntu arm Cross Checked crossgen_comparison Build and Test 
@dotnet-build-bot Test Ubuntu arm Cross Release crossgen_comparison Build and Test
@dotnet-build-bot Test Windows_NT arm Cross Checked Innerloop Build and Test 
@dotnet-build-bot Test Windows_NT arm64 Cross Checked Innerloop Build and Test

briansull commented:
Asm Diffs:

PMI Diffs for System.Private.CoreLib.dll, framework assemblies for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -7558 (-0.02% of base)
    diff is an improvement.
Top file regressions by size (bytes):
          19 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.00% of base)
          16 : System.Diagnostics.StackTrace.dasm (0.37% of base)
           7 : System.Numerics.Vectors.dasm (0.01% of base)
Top file improvements by size (bytes):
       -3463 : System.Net.Http.dasm (-0.60% of base)
       -1548 : System.Collections.dasm (-0.28% of base)
       -1136 : System.Private.Xml.dasm (-0.03% of base)
        -176 : System.Runtime.Extensions.dasm (-0.23% of base)
        -159 : System.Net.WebClient.dasm (-0.31% of base)
26 total files with size differences (23 improved, 3 regressed), 103 unchanged.
Top method regressions by size (bytes):
          85 ( 4.14% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - OverloadResolution:CollectOverloadedCandidate(ref,struct,struct,struct,struct,ref,ref,bool,ref,byref,byref)
          76 ( 2.93% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Binder:AnalyzeAnonymousFunction(ref,ref):ref:this
          42 ( 1.26% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - OverloadResolution:OperatorInvocationOverloadResolution(ref,ref,ref,ref,bool,bool,byref):struct
          20 ( 3.76% of base) : Microsoft.CodeAnalysis.CSharp.dasm - OverloadResolution:UnaryOperatorEasyOut(int,ref,ref):this
          16 ( 2.07% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - DocumentationCommentWalker:ProcessErrorLocations(ref,struct,ref,struct):this
Top method improvements by size (bytes):
       -1011 (-6.67% of base) : System.Net.Http.dasm - <SendAsyncCore>d__54:MoveNext():this
        -512 (-10.75% of base) : System.Private.Xml.dasm - <_ParseTextAsync>d__538:MoveNext():this
        -276 (-7.70% of base) : System.Net.Http.dasm - <ConnectAsync>d__52:MoveNext():this
        -258 (-18.42% of base) : System.Collections.dasm - KeyCollection:GetEnumerator():struct:this (7 methods)
        -258 (-16.58% of base) : System.Collections.dasm - KeyCollection:System.Collections.Generic.IEnumerable<TKey>.GetEnumerator():ref:this (7 methods)
Top method regressions by size (percentage):
           6 ( 9.52% of base) : Microsoft.CodeAnalysis.CSharp.dasm - LexicalSortKey:Compare(struct,struct):int
           8 ( 8.51% of base) : System.Numerics.Vectors.dasm - Vector2:Min(struct,struct):struct
           8 ( 8.51% of base) : System.Numerics.Vectors.dasm - Vector2:Max(struct,struct):struct
          13 ( 7.39% of base) : Microsoft.CodeAnalysis.dasm - SubText:.ctor(ref,struct):this
          16 ( 5.35% of base) : System.Private.Xml.dasm - GenerateHelper:EmitUnconditionalBranch(struct,struct):this
Top method improvements by size (percentage):
         -27 (-26.47% of base) : Microsoft.CodeAnalysis.dasm - SyntaxTokenList:Reverse():struct:this
         -31 (-23.48% of base) : Microsoft.CodeAnalysis.dasm - SyntaxTriviaList:Reverse():struct:this
        -258 (-18.42% of base) : System.Collections.dasm - KeyCollection:GetEnumerator():struct:this (7 methods)
        -258 (-18.42% of base) : System.Collections.dasm - ValueCollection:GetEnumerator():struct:this (7 methods)
        -258 (-16.58% of base) : System.Collections.dasm - KeyCollection:System.Collections.Generic.IEnumerable<TKey>.GetEnumerator():ref:this (7 methods)
183 total methods with size differences (145 improved, 38 regressed), 190422 unchanged.

briansull commented:
/azp run coreclr-outerloop

briansull commented:
@dotnet-build-bot  Test Windows_NT arm64 Cross Checked Innerloop Build and Test

briansull commented:
@dotnet/jit-contrib PTAL

briansull commented:
@AndyAyersMS this fixes one leg of
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/837706

CarolEidt commented:
@briansull - can you characterize the kind of code that led to this issue? It seems counter-intuitive that a `GT_LCL_VAR` would have an associated offset fieldSeq even a zero-offset.

briansull commented:
This sequence require a zero-field annotation on a GT_LCL_VAR during morph before it can reach the final GT_LCL_FLD  with a two field sequence:

```
fgMorphTree BB14, stmt 32 (before)
               [001076] ----G-------              /--*  FIELD     long   m_constArray
               [001072] ----G-------              |  \--*  ADDR      byref 
               [001073] ----G-------              |     \--*  FIELD     struct blob
               [001074] ------------              |        \--*  ADDR      byref 
               [001075] ------------              |           \--*  LCL_VAR   struct V18 loc11        
               [000152] -AC---------              *  ASG       long  
               [000151] D------N----              \--*  LCL_VAR   long  (AX) V24 loc17        


fgMorphTree BB14, stmt 32 (after)
               [001075] -----+------              /--*  LCL_FLD   long   V18 loc11        [+8] Fseq[blob, m_constArray]
               [000152] -A--G+------              *  ASG       long  
               [000151] D---G+-N----              \--*  LCL_VAR   long  (AX) V24 loc17        
```

CarolEidt commented:
> This sequence requires a zero-field annotation on a GT_LCL_VAR during morph

Where is the zero-field annotation (on which node)? And what IL sequence (or prior transform) causes it?

briansull commented:
@AndyAyersMS 
With ValueNumbering, I believe that there are potential correctness issue when we drop a fieldSequence annotation.   I have fixed several issue in this area for 3.0 (one example is  #21512).  These are correctness issues because we are tracking memory stores into struct fields.   I have improved the JitDumps for this area  (see #23876) so that we can now always see when a node has a zero-field annotation. With this change I also added an important new assert that prevents us from incorrectly adding a duplicate field sequence: (like struct.Fld1.Fld1)  There were two places in the current code that were doing that.

Since I know that there are potential issues I want to fix them for 3.0, I also saw this in my work on the desktop bug 837706 the PMI run with JItStress=2 and JItStressReg where we were hitting an assert. 

I have posted the diffs that I got on Core (see comment #3 above).  Almost all were improvements,

briansull commented:
@CarolEidt 
We first morph the subtree:
```
               [001073] ----G-------              |     \--*  FIELD     struct blob
               [001074] ------------              |        \--*  ADDR      byref 
               [001075] ------------              |           \--*  LCL_VAR   struct V18 loc11        
```
into 
`               [001075] ------------              |           \--*  LCL_VAR   struct V18 loc11      Fseq[blob]`

briansull commented on src/jit/morph.cpp:
```diff
@@ -6607,19 +6599,44 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
+        }
+        assert((addr->TypeGet() == TYP_BYREF) || (addr->TypeGet() == TYP_I_IMPL) || (addr->OperGet() == GT_LCL_FLD) ||
+               (addr->OperGet() == GT_LCL_VAR));
+
```
Added

CarolEidt replied:
This is still confusing to me. The case you show above has the op1 as an `ADDR` in both cases, and both are `TYP_BYREF`. In what case would we have something that wasn't a `TYP_BYREF` or a `TYP_I_IMPL`? And how would it be a `GT_LCL_FLD` or a `GT_LCL_VAR`.

briansull commented on src/jit/morph.cpp:
```diff
@@ -18692,66 +18722,115 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
-    assert(op1->TypeGet() == TYP_BYREF || op1->TypeGet() == TYP_I_IMPL || op1->TypeGet() == TYP_REF);
+    assert(addr->TypeGet() == TYP_BYREF || addr->TypeGet() == TYP_I_IMPL || addr->OperGet() == GT_LCL_FLD ||
+           addr->OperGet() == GT_LCL_VAR);
+    FieldSeqNode* fieldSeqUpdate   = fieldSeqZero;
```
Added

briansull commented on src/jit/morph.cpp:
```diff
@@ -18692,66 +18722,115 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 };
 
-void Compiler::fgAddFieldSeqForZeroOffset(GenTree* op1, FieldSeqNode* fieldSeq)
+void Compiler::fgAddFieldSeqForZeroOffset(GenTree* addr, FieldSeqNode* fieldSeqZero)
```
Added method header comment

briansull commented:
The tree was produced from this IL during inlining:

```
*************** In impImport() for System.Reflection.ConstArray:get_Signature():long:this

impImportBlockPending for BB70

Importing BB70 (PC=000) of 'System.Reflection.ConstArray:get_Signature():long:this'
    [ 0]   0 (0x000) ldarg.0
    [ 1]   1 (0x001) ldfld 04001C85
    [ 1]   6 (0x006) ret

    Inlinee Return expression (before normalization)  =>
               [001076] ----G-------              *  FIELD     long   m_constArray
               [001072] ----G-------              \--*  ADDR      byref 
               [001073] ----G-------                 \--*  FIELD     struct blob
               [001074] ------------                    \--*  ADDR      byref 
               [001075] ------------                       \--*  LCL_VAR   struct V18 loc11        
```

AndyAyersMS commented:
> I believe that there are potential correctness issue when we drop a fieldSequence 

I am willing to believe this too, but it would be good to understand exactly how it leads to bugs. Presumably the diffs you are showing above are from better codegen, not from fixing buggy codegen?

Does the desktop assert case lead to bad codegen in release?  

How do we know we're still not missing cases? The field sequence and zero offset field map maintenance seems easy to overlook.

Finally, is this related in any way to #22900?

briansull commented:
Yes the diffs are improvement when we are able to eliminate struct copies:

```
; Assembly listing for method ValueTask`1:ConfigureAwait(bool):struct:this
<;  V03 loc0         [V03,T07] (  2,  2   )  struct (48) [rsp+0x60]   do-not-enreg[SB] must-init ld-addr-op
>;* V03 loc0         [V03    ] (  0,  0   )  struct (48) zero-ref    do-not-enreg[SB] ld-addr-op

G_M58125_IG05:
<       vmovdqu  xmm0, qword ptr [rsp+30H]
<       vmovdqu  qword ptr [rsp+60H], xmm0
<       vmovdqu  xmm0, qword ptr [rsp+40H]
<       vmovdqu  qword ptr [rsp+70H], xmm0
<       vmovdqu  xmm0, qword ptr [rsp+50H]
<       vmovdqu  qword ptr [rsp+80H], xmm0
<
<G_M58125_IG06:
<       vmovdqu  xmm0, qword ptr [rsp+60H]
<       vmovdqu  qword ptr [rsp], xmm0
<       vmovdqu  xmm0, qword ptr [rsp+70H]
<       vmovdqu  qword ptr [rsp+10H], xmm0
<       vmovdqu  xmm0, qword ptr [rsp+80H]
<       vmovdqu  qword ptr [rsp+20H], xmm0

>; Total bytes of code 209, prolog size 32 for method ValueTask`1:ConfigureAwait(bool):struct:this
<; Total bytes of code 124, prolog size 29 for method ValueTask`1:ConfigureAwait(bool):struct:this
```

briansull commented:
Original AsmDiffs were incorrect.

Here are the actual AsmDiff for this change:

```
Summary:
(Lower is better)
Total bytes of diff: -48 (0.00% of base)
    diff is an improvement.
Top file regressions by size (bytes):
           4 : System.Private.Xml.dasm (0.00% of base)
           3 : System.Private.DataContractSerialization.dasm (0.00% of base)
Top file improvements by size (bytes):
         -55 : System.Data.Common.dasm (0.00% of base)
3 total files with size differences (1 improved, 2 regressed), 126 unchanged.
Top method regressions by size (bytes):
           4 ( 0.37% of base) : System.Private.Xml.dasm - XmlTextReaderImpl:ReadData():int:this
           3 ( 0.22% of base) : System.Private.DataContractSerialization.dasm - XmlCanonicalWriter:WriteStartAttribute(ref,int,int,ref,int,int):this
Top method improvements by size (bytes):
         -11 (-3.51% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(ubyte):int:this
         -11 (-3.51% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(short):int:this
         -11 (-3.53% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(int):int:this
         -11 (-3.38% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(double):int:this
         -11 (-3.50% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(long):int:this
Top method regressions by size (percentage):
           4 ( 0.37% of base) : System.Private.Xml.dasm - XmlTextReaderImpl:ReadData():int:this
           3 ( 0.22% of base) : System.Private.DataContractSerialization.dasm - XmlCanonicalWriter:WriteStartAttribute(ref,int,int,ref,int,int):this
Top method improvements by size (percentage):
         -11 (-3.53% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(int):int:this
         -11 (-3.51% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(ubyte):int:this
         -11 (-3.51% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(short):int:this
         -11 (-3.50% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(long):int:this
         -11 (-3.38% of base) : System.Data.Common.dasm - RBTree`1:GetNewNode(double):int:this
7 total methods with size differences (5 improved, 2 regressed), 185969 unchanged.
Completed analysis in 23.29s
```

briansull commented:
@dotnet-build-bot Test Windows_NT arm Cross Checked Innerloop Build and Test
@dotnet-build-bot Test Windows_NT arm64 Cross Checked Innerloop Build and Test

briansull commented:
@dotnet/jit-contrib PTAL

CarolEidt commented:
@briansull - could you respond to the questions that @AndyAyersMS has asked?

briansull commented:
@AndyAyersMS 

> I am willing to believe this too, but it would be good to understand exactly how it leads to bugs. 

Yes, This is an  change that improves our ability to track what is happening when we record the zero offset fields..  It adds an important assert so that we don't create incorrect duplicate field sequences  (which we are currently doing)

>Presumably the diffs you are showing above are from better codegen, not from fixing buggy codegen?

My original set of diffs were against an incorrect baseline.  There are now very few actual diffs for this change.   The few diffs that we do see involve making CSE's using  struct address calculations with zero offset fields.

>Does the desktop assert case lead to bad codegen in release?

It  resulted in an assert with the checked compiler.

> How do we know we're still not missing cases? The field sequence and zero offset field map maintenance seems easy to overlook.

Yes, I agree and that is the primary purpose of this chnage.  With this change we now print out in the JitDump every modification involving the Zero Offset tracking.  Making it much easier to follow what is happening here.   

> Finally, is this related in any way to #22900?

No, This isn't directly related to that issue, although this change will made debugging that issue and other similar issues much easier going forwards.

> How do we know we're still not missing cases?

I believe that with this refactoring we will not be dropping the Zero Field sequence information.
In particular with this last round of changes I made all of the calls to `GetZeroOffsetFieldMap()->Set()` use the method `fgAddFieldSeqForZeroOffset()`  which insures that we behave consistently.

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -13472,13 +13501,25 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
+                        }
+                        else
+                        {
+                            // Append 'fieldSeq' to the exsisting one
```
Typo "existing"

briansull replied:
fixed

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -13472,13 +13501,25 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                     {
-                        temp->ChangeOper(GT_LCL_FLD); // Note that this makes the gtFieldSeq "NotAField"...
+                        assert(temp->OperGet() == GT_LCL_VAR);
+                        temp->ChangeOper(GT_LCL_FLD); // Note that this typically makes the gtFieldSeq "NotAField"...
```
In what case would this *not* be `NotAField` - is this the zero-offset case?

briansull replied:
Yes, the code that I added with this change will set this:  (see the `compiler`.hpp diff)

```
                // Set the zeroFieldSeq in the GT_LCL_FLD node
                gtLclFld.gtFieldSeq = zeroFieldSeq;
```

CarolEidt replied:
Might be worth adding that, e.g.
```
// Note that this typically makes the gtFieldSeq "NotAField", unless we have a zero-offset FieldSeq
```
or something like that.

CarolEidt replied:
I still think it would be good to explain why it sometimes doesn't make it `NotAField`

briansull replied:
Sure I will add a comment explaining that case

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -6647,19 +6639,55 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
+#ifdef DEBUG
+        if (verbose)
+        {
+            printf("\nBefore calling fgAddFieldSeqForZeroOffset:\n");
```
This seems redundant because `fgAddFieldSeqForZeroOffset` also does a "before" dump.

briansull replied:
OK, I have removed it

briansull replied:
Actually This will print the  whole tree (and it occurs after the call to `fgMorphSmpOp`) 
It is helpful to see the final version of the tree to verify that the field sequences are correct and non of them went missing during the call to  `fgMorphSmpOp`

briansull replied:
If you want could change this to always print the final result of `Compiler::fgMorphField`.
That might be better, since the morphing of a GT_FIELD produces a new more complex tree.

CarolEidt replied:
That makes sense to me

briansull replied:
Carol after my last change the assert can now be changed to simply check that the type is BYREF or I_IMPL.

    // We expect 'addr' to be an address at this point.
    assert(addr->TypeGet() == TYP_BYREF || addr->TypeGet() == TYP_I_IMPL);

CarolEidt replied:
That's great!

briansull commented:
@dotnet-build-bot Test Windows_NT arm Cross Checked Innerloop Build and Test
@dotnet-build-bot Test Windows_NT arm64 Cross Checked Innerloop Build and Test

briansull commented:
ping @dotnet/jit-contrib PTAL

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -13478,35 +13478,38 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
-                        temp->ChangeOper(GT_LCL_FLD); // Note that this typically makes the gtFieldSeq "NotAField"...
-                        temp->AsLclFld()->gtLclOffs = (unsigned short)ival1;
+                        temp->ChangeOper(GT_LCL_FLD); // Note that this typically makes the gtFieldSeq "NotAField",
+                        // unless there is a zero filed offset associated with 'temp'.
```
The formatting here is weird, but it's not a big deal I guess.
